web/src/OverviewTable.tsx

```diff
@@ -0,0 +1,482 @@
+
+type OverviewTableProps = {
+  view: Proto.webviewView
+}
+
+type RowValues = {
+  lastDeployTime: string
+  trigger: OverviewTableTrigger
+  name: string
+  resourceTypeLabel: string
+  statusLine: OverviewTableStatus
+  podId: string
+  endpoints: UILink[]
+  triggerMode: TriggerMode
+}
+
+type OverviewTableTrigger = {
+  isBuilding: boolean
+  hasBuilt: boolean
+  hasPendingChanges: boolean
+  isQueued: boolean
+}
+
+type OverviewTableStatus = {
+  buildStatus: ResourceStatus
+  buildAlertCount: number
+  lastBuildDur: moment.Duration | null
+  runtimeStatus: ResourceStatus
+  runtimeAlertCount: number
+}
+
+const ResourceTable = styled.table`
+  margin-top: ${SizeUnit(0.5)};
+  margin-left: ${SizeUnit(1)};
+  margin-right: ${SizeUnit(1)};
+  border-collapse: collapse;
+`
+const ResourceTableHead = styled.thead`
+  background-color: ${Color.grayDarker};
+`
+const ResourceTableRow = styled.tr`
+  border-bottom: 1px solid ${Color.grayLighter};
+`
+const ResourceTableData = styled.td`
+  font-family: ${Font.monospace};
+  font-size: ${FontSize.smallest};
+  color: ${Color.gray6};
+  box-sizing: border-box;
+`
+const ResourceTableHeader = styled(ResourceTableData)`
+  color: ${Color.gray7};
+  font-size: ${FontSize.smallester};
+  padding-top: ${SizeUnit(0.5)};
+  padding-bottom: ${SizeUnit(0.5)};
+  box-sizing: border-box;
+`
+
+const ResourceTableHeaderSortTriangle = styled.div`
+  display: inline-block;
+  margin-left: ${SizeUnit(0.25)};
+  width: 0;
+  height: 0;
+  border-left: 4px solid transparent;
+  border-right: 4px solid transparent;
+  border-bottom: 6px solid ${Color.gray};
+
+  &.is-sorted-asc {
+    border-bottom: 6px solid ${Color.grayLightest};
+  }
+  &.is-sorted-desc {
+    border-bottom: 6px solid ${Color.grayLightest};
+    transform: rotate(180deg);
+  }
+`
+const ResourceName = styled.div`
+  color: ${Color.offWhite};
+  font-size: ${FontSize.small};
+  padding-top: ${SizeUnit(1 / 3)};
+  padding-bottom: ${SizeUnit(1 / 3)};
+  cursor: pointer;
+
+  &:hover {
+    text-decoration: underline;
+    text-underline-position: under;
+  }
+
+  &.has-error {
+    color: ${Color.red};
+  }
+`
+
+const Endpoint = styled.a``
+const StyledLinkSvg = styled(LinkSvg)``
+
+const DetailText = styled.span`
+  overflow: hidden;
+  text-overflow: ellipsis;
+  white-space: nowrap;
+  margin-left: 10px;
+`
+
+const PodId = styled.div`
+  display: flex;
+  align-items: center;
+`
+const PodIdInput = styled.input`
+  background-color: transparent;
+  color: ${Color.gray6};
+  font-family: inherit;
+  font-size: inherit;
+  border: 1px solid ${Color.grayDarkest};
+  border-radius: 2px;
+  padding: ${SizeUnit(0.1)} ${SizeUnit(0.2)};
+  width: 100px;
+
+  &::selection {
+    background-color: ${Color.gray};
+  }
+`
+const PodIdCopy = styled(InstrumentedButton)`
+  ${mixinResetButtonStyle};
+  padding-top: ${SizeUnit(0.5)};
+  padding: ${SizeUnit(0.25)};
+
+  svg {
+    fill: ${Color.gray6};
+  }
+`
+
+function columnDefs(): Column<RowValues>[] {
```

Comment:
would it be simpler to make these global components? e.g.,

```
function StarColumn({ row }: CellProps<RowValues>) {
  let ctx = useStarredResources()
  return (
            <TableStarResourceButton
              resourceName={row.values.name}
              analyticsName="ui.web.overviewStarButton"
              ctx={ctx}
            />
          )
        }
}

let starColumnDef = {Header: "Starred", Cell: StarColumn}
```

AFAICT, you've added two layers of indirection: a factory method layer that returns React components, and a React.memo layer that memoizes the factory methods. But the layers cancel each other out, so I'm not sure I understand what they're buying you....though I might be missing something.

Reply:
Thanks for pointing this out, @nicks. To be honest, I'm a little confused about this as well because my understanding of how and why useMemo is used here is a bit shaky. I ended up with this code structure mostly through following the sample code from React Table's [docs](https://react-table.tanstack.com/docs/quick-start), and adapting it for our needs. The library author writes that memoization is required for “referential equality from render to render,” whereas I had thought of memoization as mostly for optimization to avoid excessive re-rendering of tons of data. And when I didn't have `useMemo` in there, I kept seeing a ton of errors in the console.

If you can see a way to do this more sensibly, I'd like to pair on that. Maybe you can see a way to get around whatever's required by React Table while not getting into too much indirection.

Reply:
ya, want to pair on it this afternoon?

Reply:
pushed the changes we paired on